src/codeblockediting.js

```diff
@@ -0,0 +1,409 @@
+ */
+
+/**
+ * @module code-block/codeblockediting
+ */
+
+import Plugin from '@ckeditor/ckeditor5-core/src/plugin';
+import ShiftEnter from '@ckeditor/ckeditor5-enter/src/shiftenter';
+import CodeBlockCommand from './codeblockcommand';
+import IndentCodeBlockCommand from './indentcodeblockcommand';
+import {
+	getLocalizedLanguageDefinitions,
+	getLeadingWhiteSpaces,
+	rawSnippetTextToModelDocumentFragment
+} from './utils';
+import {
+	modelToViewCodeBlockInsertion,
+	modelToDataViewSoftBreakInsertion,
+	dataViewToModelCodeBlockInsertion
+} from './converters';
+
+const DEFAULT_ELEMENT = 'paragraph';
+
+/**
+ * The editing part of the code block feature.
+ *
+ * Introduces the `'codeBlock'` command and the `'codeBlock'` model element.
+ *
+ * @extends module:core/plugin~Plugin
+ */
+export default class CodeBlockEditing extends Plugin {
+	/**
+	 * @inheritDoc
+	 */
+	static get pluginName() {
+		return 'CodeBlockEditing';
+	}
+
+	/**
+	 * @inheritDoc
+	 */
+	static get requires() {
+		return [ ShiftEnter ];
+	}
+
+	/**
+	 * @inheritDoc
+	 */
+	constructor( editor ) {
+		super( editor );
+
+		editor.config.define( 'codeBlock', {
+			languages: [
+				{ class: 'plaintext', label: 'Plain text' },
+				{ class: 'c', label: 'C' },
+				{ class: 'cs', label: 'C#' },
+				{ class: 'cpp', label: 'C++' },
+				{ class: 'css', label: 'CSS' },
+				{ class: 'diff', label: 'Diff' },
+				{ class: 'xml', label: 'HTML/XML' },
+				{ class: 'java', label: 'Java' },
+				{ class: 'javascript', label: 'JavaScript' },
+				{ class: 'php', label: 'PHP' },
+				{ class: 'python', label: 'Python' },
+				{ class: 'ruby', label: 'Ruby' },
+				{ class: 'typescript', label: 'TypeScript' },
+			],
+
+			// A single tab.
+			indentSequence: '	'
+		} );
+	}
+
+	/**
+	 * @inheritDoc
+	 */
+	init() {
+		const editor = this.editor;
+		const schema = editor.model.schema;
+		const model = editor.model;
+
+		const localizedLanguageDefinitions = getLocalizedLanguageDefinitions( editor );
+		const languageClasses = localizedLanguageDefinitions.map( def => def.class );
+		const languageLabels = Object.assign( {}, ...localizedLanguageDefinitions.map( def => ( { [ def.class ]: def.label } ) ) );
+
+		// The main command.
+		editor.commands.add( 'codeBlock', new CodeBlockCommand( editor ) );
+
+		// Commands that change the indentation.
+		editor.commands.add( 'indentCodeBlock', new IndentCodeBlockCommand( editor, 'forward' ) );
+		editor.commands.add( 'outdentCodeBlock', new IndentCodeBlockCommand( editor, 'backward' ) );
+
+		const getCommandExecuter = commandName => {
+			return ( data, cancel ) => {
+				const command = this.editor.commands.get( commandName );
+
+				if ( command.isEnabled ) {
+					this.editor.execute( commandName );
+					cancel();
+				}
+			};
+		};
+
+		editor.keystrokes.set( 'Tab', getCommandExecuter( 'indentCodeBlock' ) );
+		editor.keystrokes.set( 'Shift+Tab', getCommandExecuter( 'outdentCodeBlock' ) );
```

Comment:
When there's nothing to outdent, the command is disabled and the keystroke works like there is no integration (starts cycling focus on the page). I wonder if this is the right thing because it surprised me a couple of times when outdenting, again and again, I found myself outside of the editor.

Reply:
I'd go with follow up for that as this is a general issue in CKEditor I think - we have such behavior in lists and maybe tables.

Reply:
Can you give me some similar scenarios? Because I wasn't able to find any.

Reply:
Indent / outdent in lists - if you'd like to go behbeyondond max level you also tab outside the eidtor.

Reply:
I'll create a follow–up.

Reply:
FYI https://github.com/ckeditor/ckeditor5/issues/5768.